log.go

```diff
@@ -20,22 +20,143 @@
 package wavelet
 
 import (
+	"encoding/hex"
+	"sync"
+	"time"
+
+	"github.com/perlin-network/wavelet/conf"
 	"github.com/perlin-network/wavelet/log"
+	"github.com/valyala/fastjson"
 )
 
-func logEventTX(event string, tx *Transaction, other ...interface{}) {
-	logger := log.TX(event)
-	log := logger.Log().
-		Hex("tx_id", tx.ID[:]).
-		Hex("sender_id", tx.Sender[:]).
-		Uint8("tag", byte(tx.Tag))
-
-	for _, o := range other {
-		switch o := o.(type) { // nolint:gocritic
-		case error:
-			log = log.Err(o)
+// CollapseResultsLogger is used to write CollapseResults to the logger's writers.
+//
+// It writes directly into the writers without going through zerolog.
+// The reason is that, zerolog will write into all the writers, even the writer's module does not match with the message's module.
+//
+// It also has a buffer to prevent blocking, as writing all the transactions in a collapse result may take sometime.
+// A collapse result may contain tens of thousands of transactions.
+type CollapseResultsLogger struct {
+	arena         *fastjson.Arena
+	mod           []byte // "tx"
+	eventApplied  []byte // "applied"
+	eventRejected []byte // "rejected"
+	timeLayout    string // "2006-01-02T15:04:05Z07:00"
+
+	bufTxID     []byte
+	bufSenderID []byte
+	bufTime     []byte
+
+	// Buffer for a batch of messages (many transactions)
+	bufBatch [][]byte
+
+	flushCh chan [][]byte
+
+	stopWg sync.WaitGroup
+	stop   chan struct{}
+	closed bool
+}
+
+func NewCollapseResultsLogger() *CollapseResultsLogger {
+	c := &CollapseResultsLogger{
+		arena:         &fastjson.Arena{},
+		mod:           []byte("tx"),
+		eventApplied:  []byte("applied"),
+		eventRejected: []byte("rejected"),
+		timeLayout:    "2006-01-02T15:04:05Z07:00",
+		bufTxID:       make([]byte, hex.EncodedLen(SizeTransactionID)),
+		bufSenderID:   make([]byte, hex.EncodedLen(SizeAccountID)),
+		bufTime:       make([]byte, 0, 64),
+
+		bufBatch: make([][]byte, 0, conf.GetBlockTXLimit()/4),
+		flushCh:  make(chan [][]byte, 1024),
```

Comment:
https://github.com/uber-go/guide/blob/master/style.md#channel-size-is-one-or-none